magellan/debugblocker/debugblocker.py

```diff
@@ -0,0 +1,633 @@
+                       cand_set, output_size):
+    total_compared_pairs = 0
+    compared_set = set()
+    l_inverted_index = {}
+    r_inverted_index = {}
+    topk_heap = []
+
+    while len(prefix_events) > 0:
+        if len(topk_heap) == output_size and\
+                        topk_heap[0][0] >= prefix_events[0][0] * -1:
+            break
+        event = hq.heappop(prefix_events)
+        table_indicator = event[1]
+        rec_idx = event[2]
+        tok_idx = event[3]
+        if table_indicator == 0:
+            token = lrecord_list[rec_idx][tok_idx]
+            if token in r_inverted_index:
+                r_records = r_inverted_index[token]
+                for r_rec_idx in r_records:
+                    pair = (rec_idx, r_rec_idx)
+                    # Skip if the pair is in the candidate set.
+                    if pair in cand_set:
+                        continue
+                    # Skip if the pair has been compared.
+                    if pair in compared_set:
+                        continue
+                    sim = jaccard_sim(
+                        set(lrecord_list[rec_idx]), set(rrecord_list[r_rec_idx]))
+                    if len(topk_heap) == output_size:
+                        hq.heappushpop(topk_heap, (sim, rec_idx, r_rec_idx))
+                    else:
+                        hq.heappush(topk_heap, (sim, rec_idx, r_rec_idx))
+
+                    total_compared_pairs += 1
+                    compared_set.add(pair)
+
+            # Update the inverted index.
+            if token not in l_inverted_index:
+                l_inverted_index[token] = set()
+            l_inverted_index[token].add(rec_idx)
+        else:
+            token = rrecord_list[rec_idx][tok_idx]
+            if token in l_inverted_index:
+                l_records = l_inverted_index[token]
+                for l_rec_idx in l_records:
+                    pair = (l_rec_idx, rec_idx)
+                    # Skip if the pair is in the candidate set.
+                    if pair in cand_set:
+                        continue
+                    # Skip if the pair has been compared.
+                    if pair in compared_set:
+                        continue
+                    sim = jaccard_sim(
+                        set(lrecord_list[l_rec_idx]), set(rrecord_list[rec_idx]))
+                    if len(topk_heap) == output_size:
+                        hq.heappushpop(topk_heap, (sim, l_rec_idx, rec_idx))
+                    else:
+                        hq.heappush(topk_heap, (sim, l_rec_idx, rec_idx))
+
+                    total_compared_pairs += 1
+                    compared_set.add(pair)
+
+            # Update the inverted index.
+            if token not in r_inverted_index:
+                r_inverted_index[token] = set()
+            r_inverted_index[token].add(rec_idx)
+
+    return topk_heap
+
+
+# Calculate the token-based Jaccard similarity of two string sets.
```

Comment:
Discussed about this. Currently stick to the original implementation for efficiency. Maybe we should switch to py_stringmatching later.